jobs/e2e-test.groovy

```diff
@@ -302,6 +305,19 @@ def runTestServer() {
       // before it does so, so it doesn't read "old" data.
       sh("rm -rf genfiles/test-reports");
 
+      tests = sh(script: "testing/runtests_server.py -n ${exec.shellEscapeList(runSmokeTestsArgs)} . ", returnStdout: true);
+      // The runtests_server.py script with -n outputs all tests to run of 
+      // various types. We only need to worry about smoke tests, which are 
+      // also the last section, so grab everything after the header. If it's 
+      // empty, there are no tests to run.
+      e2eTests = tests.substring(tests.lastIndexOf("SMOKE TESTS:") + 13).trim()
+      echo("e2eTests: ${e2eTests}")
+      if (e2eTests.isAllWhitespace()) {
+         echo("No E2E Tests to run!")
+         skipTestAnalysis = true
+         return
```

Comment:
I think the run will be more effective if you do `throw new TestsAreDone()` here?  Likewise below.  It's unclear to me if this is still causing all the clients to start up or not.  Honestly, I am surprised the code you have is working as-is.  What is stopping the clients from starting up, trying to talk to the server, and being upset they can't?

I think at least you need to be doing this before you set TEST_SERVER_URL, not after.

It's hard for me to tell what testing you actually did.  It would be great if your test plan were a recipe that other people could follow to confirm that things work correctly.  Something like:
```
I started a jenkins e2e job with the old code and set TEST_TO_RUN to <something> and TESTS_TO_SKIP to <same thing>.  I ran the job and saw it errored out.  Then I clicked on "replay" and replaced the main e2e code with the code in this commit, and ran it again, and this time it didn't error out.
```
It would be good to verify that both runs take about the same amount of time, as well!  Or that the new code is faster. :-)

Reply:
> I think the run will be more effective if you do `throw new TestsAreDone()` here? Likewise below. It's unclear to me if this is still causing all the clients to start up or not. Honestly, I am surprised the code you have is working as-is. What is stopping the clients from starting up, trying to talk to the server, and being upset they can't?

The clients do fail (but don't fail the job), and I was trying to find a way to make them quit more quickly, so that's great if that works that simply. I was unaware of that. I'm moderately concerned it will be difficult to achieve the desired path without a lot of testing given the potential threading and race-condition implications, but it's probably worth a quick attempt.


> I think at least you need to be doing this before you set TEST_SERVER_URL, not after.

Oh, because setting that unblocks the workers. Good point, that's an. easy fix. Sorry I didn't see that comment. We could probably similarly exit the test workers before starting them if the `skipTestAnalysis` is false (though probably want to rename it rather than overload it like that).


> It's hard for me to tell what testing you actually did. It would be great if your test plan were a recipe that other people could follow to confirm that things work correctly. Something like:
> 
> > I started a jenkins e2e job with the old code and set TEST_TO_RUN to `<something>` and TESTS_TO_SKIP to `<same thing>`.  I ran the job and saw it errored out.  Then I clicked on "replay" and replaced the main e2e code with the code in this commit, and ran it again, and this time it didn't error out.

It's not clear to me how what you're asking for is different than the testing plan I've already provided, which includes both links to runs where this was tested, as well as instructions for how to replicate each scenario. I suppose I didn't call out the fact that the first run of each do-it-yourself option will result in a failure?


> It would be good to verify that both runs take about the same amount of time, as well! Or that the new code is faster. :-)

That's reasonable, though I will point out as we have found many times before that benchmarking jenkins on a job-by-job basis is unreliable, since so much of the processes have to wait on external factors. Q&D analysis (based on clicking the available passed jobs in jenkins):

Unit test take 10-35 minutes to run when they pass, with an average that looks to be in the 14-17 minute range. The test case where I ran the full suite of tests above with the code here [webapp-test 113728](https://jenkins.khanacademy.org/job/deploy/job/webapp-test/113728/) ran in 13 minutes, so just a hair under the average - not surprising, since it was doing exactly the same work other than a very fast bit of computation. The case where there are no unit tests [webapp-test 113727](https://jenkins.khanacademy.org/job/deploy/job/webapp-test/113727/) ran in 1 minute 7 seconds, so quite a bit faster (it's previous failures were similarly fast, fwiw).

E2e testing takes between 4 and 20 minutes based on a similar quick analysis. The test run above which had one test case,  [e2e-test 31199](https://jenkins.khanacademy.org/job/deploy/job/e2e-test/31199/) ran in 4m32s, so on par with the fastest runs. In only had one test to run, so that may not be a perfect picture, I'll make a run with this code that doesn't skip anything and see how it performs. The case where there were zero e2e tests to run [e2e-test 31202](https://jenkins.khanacademy.org/job/deploy/job/e2e-test/31202/) ran in 1m01s, so quite quickly.

That last number gives me optimism that the extra call to `testing/runtests_server.py -n` isn't costing all. that much time, though it's certainly possible it works significantly faster when there are no tests to run...

Reply:
No, it takes a constant amount of time to run because it has to read the whole filesystem.

When it comes to timing, I think what you'll want to do is to go to the "pipeline steps" page and see how long individual pipeline steps take.  Like see how long the clients are running for before and after your change.  And then you can see how long the extra runtests_server call is taking directly.

Reply:
> It's not clear to me how what you're asking for is different than the testing plan I've already provided, which includes both links to runs where this was tested, as well as instructions for how to replicate each scenario. I suppose I didn't call out the fact that the first run of each do-it-yourself option will result in a failure?

I think the key is to write a test plan that can be followed by someone who doesn't really understand the system.  So it needs to be very precise: not just "run the job" but "visit https://jenkins.khanacademy.org/job/deploy/job/e2e-test" and click on the "build now" button, and put "<whatever>" in the "version" field and click "build".  (And I usually frame it not as commands you should run but commands I ran: "I visited https://... and clicked "build"", though practices here differ.)  I think you pointed to a generic "how to use jenkins" page as a substitute for that, but that's not as good a substitute, since that page covers a lot of things that *aren't* appropriate for testing this change.

As a specific example, I don't understand from your test plan how you ran your new code jenkins-job, rather than the existing code.

Reply:
Oh, I see! I provided the link to the wiki article that describes how to run the steps in the message but that might not have been obvious since it was just labelled 'here'. I can add a better description.